src/txn/txn_ckpt.c

```diff
@@ -627,6 +629,22 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
 
     __wt_writeunlock(session, &txn_global->rwlock);
 
+    /*
+     * Open an internal session to allocate a transaction id that is used for removing the history
+     * store entries due to a prepared rollback occurs in parallel to the checkpoint. Make sure that
+     * this transaction id is published before the checkpoint acquires its snapshot.
+     */
+    WT_RET(__wt_open_internal_session(conn, "checkpoint internal", true, session->flags,
+      session->lock_flags, &ckpt_internal_session));
+    WT_ERR(__wt_txn_begin(ckpt_internal_session, NULL));
+    WT_ERR(__wt_txn_id_check(ckpt_internal_session));
+    txn_global->checkpoint_invisible_txn_id = ckpt_internal_session->txn->id;
```

Comment:
Is it safe to do this outside the read write lock? I assume we can't do it inside the lock as one of the txn calls will try and lock that structure. We don't mark the checkpoint_invisible_txn_id as volatile either but I think that's okay as it's only read once in the function that uses it.

Reply:
The id allocation is performed outside of the lock always. I changed the variable as volatile to make sure that ordering shouldn't cause the problem.